dttools/src/path_disk_size_info.c

```diff
@@ -47,102 +42,117 @@ int path_disk_size_info_get(const char *path, int64_t *measured_size, int64_t *n
 }
 
 int path_disk_size_info_get_r(const char *path, int64_t max_secs, struct path_disk_size_info **state) {
-	int64_t start_time = time(0);
-	int result = 0;
 
 	if(!*state) {
 		/* if state is null, there is no state, and path is the root of the measurement. */
 		*state = calloc(1, sizeof(struct path_disk_size_info));
 	}
-
 	struct path_disk_size_info *s = *state;     /* shortcut for *state, so we do not need to type (*state)->... */
 
+	int64_t start_time = time(0);
+	int result = 0;
+	
 	/* if no current_dirs, we begin a new measurement. */
 	if(!s->current_dirs) {
 		s->complete_measurement = 0;
+		s->current_dirs 		= list_create();
+		s->size_so_far 			= 0;
+		s->count_so_far 		= 0;
+		list_push_tail(s->current_dirs, (void*) xxstrdup(path));
+	}
 
-		struct DIR_with_name *here = malloc(sizeof(struct DIR_with_name));
+	/* Start processing unseen directories.
+	 * Each loop counts the current directory and all files immediately except 
+	 * files that are directories. Directory entries are added to a list to be 
+	 * counted later.
+	 */
+	char* tail_name;
+	struct dirent *entry;
+	struct stat file_info;
+	DIR* tail_dir;
+	int timeout = 0;
+	while((tail_name = list_pop_tail(s->current_dirs))) {
+		tail_dir = opendir(tail_name);
```

Comment:
I'm not sure about this one. The purpose of keeping the open dirs with DIR_with_name is so that the measurement can be interrupted on a timeout and then resumed given the state. If the opendir is called here, then either the previous progress is lost, or we can't interrupt the call.

I'd suggest a minimal change to the previous version.  Keep the DIR_with_name struct, and only open the dir when popping the struct from the list.

Reply:
Ah, that is a better approach, thanks!